imagebuildah/cache/local_layer_provider.go

```diff
@@ -0,0 +1,135 @@
+package cache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/containers/image/v5/manifest"
+	is "github.com/containers/image/v5/storage"
+	"github.com/containers/storage"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// LocalLayerProvider is an implementation of the LayerProvider that
+type LocalLayerProvider struct {
+	store storage.Store
+
+	localLayers     map[string]string
+	localLayersLock sync.Mutex
+}
+
+// NewLocalLayerProvider creates a new instance of the LocalLayerProvider.
+func NewLocalLayerProvider(store storage.Store) LocalLayerProvider {
+	return LocalLayerProvider{
+		store:           store,
+		localLayers:     make(map[string]string),
+		localLayersLock: sync.Mutex{},
+	}
+}
+
+// CalculateLocalLayerKey returns the layer key for image in the local store
+func CalculateLocalLayerKey(manifestType string, parentLayerID string, history []v1.History, digests []digest.Digest) string {
+	lastElement := len(history) - 1
+	buildAddsLayer := !history[lastElement].EmptyLayer
+	createdBy := history[lastElement].CreatedBy
+	return CalculateBuildLayerKey(manifestType, buildAddsLayer, parentLayerID, createdBy, history[:lastElement], digests)
+}
+
+// getImageTypeAndHistoryAndDiffIDs returns the manifest type, history, and diff IDs list of imageID.
+func (llp *LocalLayerProvider) getImageTypeAndHistoryAndDiffIDs(ctx context.Context, imageID string) (string, []v1.History, []digest.Digest, error) {
+	imageRef, err := is.Transport.ParseStoreReference(llp.store, "@"+imageID)
+	if err != nil {
+		return "", nil, nil, errors.Wrapf(err, "error getting image reference %q", imageID)
+	}
+	ref, err := imageRef.NewImage(ctx, nil)
+	if err != nil {
+		return "", nil, nil, errors.Wrapf(err, "error creating new image from reference to image %q", imageID)
+	}
+	defer ref.Close()
+	oci, err := ref.OCIConfig(ctx)
+	if err != nil {
+		return "", nil, nil, errors.Wrapf(err, "error getting possibly-converted OCI config of image %q", imageID)
+	}
+	manifestBytes, manifestFormat, err := ref.Manifest(ctx)
+	if err != nil {
+		return "", nil, nil, errors.Wrapf(err, "error getting manifest of image %q", imageID)
+	}
+	if manifestFormat == "" && len(manifestBytes) > 0 {
+		manifestFormat = manifest.GuessMIMEType(manifestBytes)
+	}
+	return manifestFormat, oci.History, oci.RootFS.DiffIDs, nil
+}
+
+// PopulateLayer scans the local images and adds them to the map.
+func (llp *LocalLayerProvider) PopulateLayer(ctx context.Context, topLayer string) error {
+	llp.localLayersLock.Lock()
+	defer llp.localLayersLock.Unlock()
+
+	// Get the list of images available in the image store
+	images, err := llp.store.Images()
+	if err != nil {
+		return errors.Wrap(err, "error getting image list from store")
```

Comment:
Please do not start error messages with "error"

Reply:
resolved